lib/firmware/cosmos.c

```diff
@@ -0,0 +1,307 @@
+#include "keepkey/firmware/cosmos.h"
+
+#include "keepkey/board/confirm_sm.h"
+#include "keepkey/board/util.h"
+#include "keepkey/firmware/home_sm.h"
+#include "keepkey/firmware/storage.h"
+#include "trezor/crypto/ecdsa.h"
+#include "trezor/crypto/memzero.h"
+#include "trezor/crypto/segwit_addr.h"
+#include <stdbool.h>
+#include <time.h>
+
+static CONFIDENTIAL HDNode node;
+static SHA256_CTX ctx;
+static bool has_message;
+static bool initialized;
+static uint32_t address_n[8];
+static size_t address_n_count;
+static uint32_t msgs_remaining;
+static uint64_t sequence;
+
+bool cosmos_path_mismatched(const CoinType *_coin,
+                            const uint32_t *_address_n,
+                            const uint32_t _address_n_count)
+{
+    // m/44' : BIP44-like path
+    // m / purpose' / bip44_account_path' / account' / x / y
+    bool mismatch = false;
+    mismatch |= _address_n_count != 5;
+    mismatch |= _address_n_count > 0 && (_address_n[0] != (0x80000000 + 44));
+    mismatch |= _address_n_count > 1 && (_address_n[1] != _coin->bip44_account_path);
+    mismatch |= _address_n_count > 2 && (_address_n[2] & 0x80000000) == 0;
+    mismatch |= _address_n_count > 3 && (_address_n[3] & 0x80000000) != 0;
+    mismatch |= _address_n_count > 4 && (_address_n[4] & 0x80000000) != 0;
+    return mismatch;
+}
+
+/*
+ * Gets the address
+ *
+ * public_key: 33 byte compressed secp256k1 key
+ * address: output buffer
+ *
+ * returns true if successful
+ */
+bool cosmos_getAddress(const HDNode *_node, char *address)
+{
+    uint8_t hash160Buf[RIPEMD160_DIGEST_LENGTH];
+    ecdsa_get_pubkeyhash(_node->public_key, HASHER_SHA2_RIPEMD, hash160Buf);
+
+    uint8_t fiveBitExpanded[RIPEMD160_DIGEST_LENGTH * 8 / 5];
+    size_t len = 0;
+    convert_bits(fiveBitExpanded, &len, 5, hash160Buf, 20, 8, 1);
+    // bech32encode
+    return bech32_encode(address, "cosmos", fiveBitExpanded, len) == 1;
+}
+
+// Each segment guaranteed to be less than or equal to 64 bytes
+// 19 + ^20 + 14 = ^53
+#define SIGNING_TEMPLATE_START_SEG1 "{\"account_number\":\"%" PRIu64 "\",\"chain_id\":\""
```

Comment:
I think these should all be inlined where they're used. The macros aren't buying us much, and they arguably make it less clear what's going on.

Reply:
I'm fine doing that, but the reason we went this route is because the format strings have to either be literals or macros. If we inline them it adds a bit of noise to the line and we had opted to extract that noise to a single place. I agree that the *_SEG# is somewhat arbitrary and for that reason it can be very unclear. However, it was important that all of the strings with their maximum expansions for the wildcards fell below the buffer size.